### PR TITLE
7706 add close button to flatpickr

### LINKED ISSUE #7706: Add a "Close/Confirm" button into the date(time)picker
Despite that the issue #7697 should be fixed for the order cycles list in the backoffice, we are wondering about the opportunity to add a Close/Confirm button. 
This could make event cleaner to the user that the date(time)picker should be closed before saving its value.
The issue is here to discuss about and then apply the chosen fix.

Reference: 
Capture d’écran 2021-05-27 à 11 54 39

_Originally posted by @jibees in https://github.com/openfoodfoundation/openfoodnetwork/issues/7689#issuecomment-849501296_

### PR DESCRIPTION
**Pull request by KombiCode**
#### What? Why?
Closes #7706

After discussions it was decided to add a 'Close' button which just closes the popover.
This should make things clearer for user on how to close

#### What should we test?

The new button has been added into datetimepicker and datepicker, so one need to test both.

As an admin go to eg Order Cycles, click on one datetime area to open the calendar, you should see a Close button on the right like illustrated in this screenshot :
Capture d’écran 2021-10-27 à 11 16 22
Clicking on it should close the popover
( A test has been added for this case in list_spec.rb in context 'using datetimepickers')

Similarly select Orders from the admin dashboard and click on eg Start date area, you should also see the Close button on the right like illustrated in this screenshot :
Capture d’écran 2021-10-27 à 11 15 57
Clicking on it should close the popover
(A test has been added for this case in orders_spec.rb in context "with a complete order"

#### Release notes

Calendars now have a Close button to clearly indicate to the user a way to close the popover 

Changelog Category: User facing changes

### CODE REVIEW

File: spec/system/admin/order_cycles/list_spec.rb [hunk 1]
```diff
@@ -147,6 +147,32 @@
...
+        # Opens a datetimepicker
+        within("tr.order-cycle-#{oc_pt.id}") do
+          find('input.datetimepicker', match: :first).click
+        end
+
+        # Sets the value to test_value then looks for the close button and click it
+        within(".flatpickr-calendar.open") do
+          expect(page).to have_selector '.shortcut-buttons-flatpickr-buttons'
+          select_datetime_from_datepicker test_value
+          find(".shortcut-buttons-flatpickr-buttons > button:nth-last-child(1)").click
```
Comment (mkllnk) [inline spec/system/admin/order_cycles/list_spec.rb:164]: It's better to test for what the user sees:
```suggestion
          find("button", text: "CLOSE").click
```

File: spec/support/features/datepicker_helper.rb [hunk 2]
```diff
@@ -4,7 +4,8 @@ module Features
   module DatepickerHelper
     def choose_today_from_datepicker
       within(".flatpickr-calendar.open") do
-        find('.shortcut-buttons-flatpickr-button').click
+        # Now that we have added the Close button, find the first button to avoid any ambiguous match
+        find('.shortcut-buttons-flatpickr-buttons > button:nth-child(1)').click
```
Comment (mkllnk) [inline spec/support/features/datepicker_helper.rb:8]: Does this work?
```suggestion
        find("button", text: "TODAY").click
```

File: spec/system/flatpickr_spec.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      open_datepicker('#q_completed_at_gteq')       
+      # Looks for the close button and click it
+      within(".flatpickr-calendar.open") do
+        expect(page).to have_selector '.shortcut-buttons-flatpickr-buttons'
+        find(".shortcut-buttons-flatpickr-buttons > button:nth-last-child(1)").click
+      end
+      # Should no more have opened flatpickr
+      expect(page).not_to have_selector '.flatpickr-calendar.open'
+    end
+    it "opens the datepicker and sets date to today" do
```
Comment (mkllnk) [inline spec/system/flatpickr_spec.rb:21]: We usually have empty lines in between examples.
```suggestion
    end

    it "opens the datepicker and sets date to today" do
```

File: spec/system/flatpickr_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      login_as_admin_and_visit 'admin/orders'
+      open_datepicker('#q_completed_at_gteq')
+      find("#admin-menu").click       
+      # Should no more have opened flatpickr
+      expect(page).not_to have_selector '.flatpickr-calendar.open'
+    end
+  end
+  
+private
+  def open_datepicker(field)
```
Comment (mkllnk) [inline spec/system/flatpickr_spec.rb:37]:
```suggestion
  private

  def open_datepicker(field)
```

**Review by mkllnk — CHANGES_REQUESTED** [review]
Great work!

I just added some suggestions to make it easier to read.

**Comment by KombiCode** [comment]
> Great work!
> 
> I just added some suggestions to make it easier to read.

Thanks Maikel. I'm doing the requested changes and will push them asap.

**Review by mkllnk — APPROVED** [review]
:tada:

**Comment by filipefurtad0** [comment]
Hey @KombiCode ,

I think this is first contribution - welcome! :tada:

This looks great. I've checked several pages:

/admin/order_cycles/
/admin/order_cycles/[ORDER_ID]/edit
/admin/subscriptions/[subscription_id]/edit
/admin/orders/bulk_management
/admin/orders?q[s]=completed_at+desc#


and reports:

Orders And Distributors
Bulk Co-Op
Payment Reports
Orders & Fulfillment Reports
Enterprise Fee Summary
Order Cycle Management
Sales Tax
Xero Invoices
Packing Reports

Posting only these two as examples:

**Orders page - date-picker**
image

**Order Cycles - date-time-picker**
image

This introduces no changes on mobile - this is expected:

Android 12, Chrome
image

Android 12, Firefox 
image

iPhone / Safari
image


**Summary**

We now see the `Close` button now appears; it behaves similarly as to clicking away from the calendar area, before or after editing dates. 

Awesome work - thanks for adding the tests too :tada: 
Good to go.

**Comment by KombiCode** [comment]
> Hey @KombiCode ,
> 
> I think this is first contribution - welcome! 🎉
> 
> This looks great. I've checked several pages:
> 
> /admin/order_cycles/ /admin/order_cycles/[ORDER_ID]/edit /admin/subscriptions/[subscription_id]/edit /admin/orders/bulk_management /admin/orders?q[s]=completed_at+desc#
> 
> and reports:
> 
> Orders And Distributors Bulk Co-Op Payment Reports Orders & Fulfillment Reports Enterprise Fee Summary Order Cycle Management Sales Tax Xero Invoices Packing Reports
> 
> Posting only these two as examples:
> 
> **Orders page - date-picker** image
> 
> **Order Cycles - date-time-picker** image
> 
> This introduces no changes on mobile - this is expected:
> 
> Android 12, Chrome image
> 
> Android 12, Firefox image
> 
> iPhone / Safari image
> 
> **Summary**
> 
> We now see the `Close` button now appears; it behaves similarly as to clicking away from the calendar area, before or after editing dates, as mentioned [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/7706#issuecomment-951704505).
> 
> Awesome work - thanks for adding the tests too 🎉 Good to go.

Hi @filipefurtad0,

Thanks for your comments and checks.
Indeed this is my first contribution 😅, will start another one soon.

**Pull request merged**